### PR TITLE
Rework cidata

### PR DESCRIPTION
**Pull request by jgrassler**
This pull request reworks cidata to. Main changes:

* Fix exit code handling
* Implement the new contract for read()
* Append STDOUT/STDERR to error message on non-zero exit code.

Plus a bunch of small fixes that occured to me on the way.

### CODE REVIEW

File: cidata.py [hunk 1]
```diff
@@ -561,74 +568,86 @@ def read(
...
             username='robot',
         )
     except CouldNotConnectException:
-        return False, data_dict, messages[3221]
+        retval = False
+        message_list.append(messages[3221])
 
-    if read_metadata.exit_code != SUCCESS_CODE:
-        return False, data_dict, messages[3222] + f'{read_metadata.exit_code}s.'
+    if exit_code != SUCCESS_CODE:
```
Comment (maria-walsh) [inline cidata.py:595]: `exit_code` may be undefined if the rcc.comms_ssh() could not connect.

File: cidata.py [hunk 2]
```diff
@@ -561,74 +568,86 @@ def read(
...
+        retval = False
+        message_list.append(messages[3216])
     elif podnet_a_enabled is False and podnet_b_enabled is False:
-        return False, data_dict, messages[3217]
+        retval = False
+        message_list.append(messages[3217])
     else:
-        return False, data_dict, messages[3218]
+        retval = False
+        message_list.append(messages[3218])
```
Comment (maria-walsh) [inline cidata.py:579]: If any "message" has been raised so far, the remainder of this read function will also fail. 

I think if there is an issue with any of the values in config.json we should return here and only continue if the config.json is valid.

File: cidata.py [hunk 3]
```diff
@@ -561,74 +568,86 @@ def read(
...
 
     # define payloads
     read_userdata_payload = f'cat {domain_path}/userdata'
     read_metadata_payload = f'cat {domain_path}/metadata'
 
     # call rcc comms_ssh for metadata retrieval from enabled PodNet
     try:
-        read_metadata, stdout, stderr = comms_ssh(
+        exit_code, stdout, stderr = comms_ssh(
             host_ip=enabled,
```
Comment (maria-walsh) [inline cidata.py:587]: `enabled` may be undefined if the config.json was invalid.

**Review by maria-walsh — CHANGES_REQUESTED** [review]

File: cidata.py (hunk 2 above)
Comment (jgrassler) [reply cidata.py:578]: Makes sense, yeah.

File: cidata.py (hunk 1 above)
Comment (jgrassler) [reply cidata.py]: Fixed.

File: cidata.py (hunk 3 above)
Comment (jgrassler) [reply cidata.py:586]: Yeah. Definitely a good idea to return early on config.json errors.

File: cidata.py [hunk 4]
```diff
@@ -561,74 +568,92 @@ def read(
...
+
+    if retval == False:
+        return retval, data_dict, message_list
 
     # define payloads
     read_userdata_payload = f'cat {domain_path}/userdata'
     read_metadata_payload = f'cat {domain_path}/metadata'
 
     # call rcc comms_ssh for metadata retrieval from enabled PodNet
     try:
```
Comment (maria-walsh) [inline cidata.py:587]:
```python
try:
    exit_code, stdout, stderr = comms_ssh(
        host_ip=enabled,
        payload=read_metadata_payload,
        username='robot',
    )
    if exit_code != SUCCESS_CODE:
        retval = False
        message_list.append(messages[3222] + f'{exit_code}s.\nSTDOUT: {stdout}\nSTDERR: {stderr}')
except CouldNotConnectException:
    retval = False
    message_list.append(messages[3221])
```

Would this be easier to manage as stdout and stderr will also be undefined?

**Pull request merged**